### PR TITLE
Added rop support for branch slots

### LINKED ISSUE #2090: MIPS rop doesn't take into consideration pipelined instruction after jmp
mips has delay slot stuff

http://en.wikipedia.org/wiki/Delay_slot

### PR DESCRIPTION
**Pull request by jcorina**
ROP chains now have branch slots included in them.

Closes #2090

### CODE REVIEW

File: libr/core/cmd_search.c [hunk 1]
```diff
@@ -736,6 +744,30 @@ static void print_rop (RCore *core, RList *hitlist, char mode, int *json_first)
 	if (mode != 'j') r_cons_newline ();
 }
 
+
+static int has_branch_delay_slot(const char* arch) {
+	
+	if (strncmp(arch,"mips",4))
```
Comment (crowell) [inline libr/core/cmd_search.c]: strncmp is done wrong, it returns 0 on equality

also, can you collapse this up a bit? not as many empty lines

File: libr/core/cmd_search.c [hunk 2]
```diff
@@ -747,6 +779,7 @@ static int r_core_search_rop(RCore *core, ut64 from, ut64 to, int opt, const cha
 	boolt json_first = 1;
 	const char *smode = r_config_get (core->config, "search.in");
 	const char *arch = r_config_get (core->config, "asm.arch");
+	int branch_delay = has_branch_delay_slot(arch);	//Whether or not we have branch delay slots
```
Comment (crowell) [inline libr/core/cmd_search.c]: space after comment

File: libr/core/cmd_search.c [hunk 3]
```diff
@@ -851,7 +884,12 @@ static int r_core_search_rop(RCore *core, ut64 from, ut64 to, int opt, const cha
 					// limit number of high level rop gadget results
 					break;
 				}
-				r_list_append (end_list, (void*)(intptr_t)i);
+
+				//if this arch has branch delay slots, add the next instr as well
```
Comment (crowell) [inline libr/core/cmd_search.c]: space after //

File: libr/core/cmd_search.c [hunk 4]
```diff
@@ -851,7 +884,12 @@ static int r_core_search_rop(RCore *core, ut64 from, ut64 to, int opt, const cha
...
+				if ( branch_delay && (end_gadget.type != R_ANAL_OP_TYPE_TRAP) )
```
Comment (crowell) [inline libr/core/cmd_search.c]: `if (branch` spacing

File: libr/core/cmd_search.c [hunk 5]
```diff
@@ -658,6 +658,14 @@ static RList* construct_rop_gadget(RCore *core, ut64 addr, ut8 *buf, int idx,
 		r_list_append (badstart, p);
 	}
 	r_list_free (localbadstart);
+	//if our arch has bds then we better be including them
+	if (branch_delay) {
```
Comment (crowell) [inline libr/core/cmd_search.c]: `if (branch_delay && r_list_length(hitlist) < 2)` would be more readable imo

**Comment by radare** [comment]
The delay slot must is provided by the anal plugin and taking the value from there. On sparc there's a 32 byte slot.

This delay information can be taken by analyzing the instruction in op.delay.

**Comment by jcorina** [comment]
Not very familiar with the codebase yet so let me know if I made any mistakes/caused any issues.

File: libr/core/cmd_search.c [hunk 6]
```diff
@@ -748,7 +761,7 @@ static int r_core_search_rop(RCore *core, ut64 from, ut64 to, int opt, const cha
 	const char *smode = r_config_get (core->config, "search.in");
 	const char *arch = r_config_get (core->config, "asm.arch");
 	RList/*<RRegex>*/ *rx_list = NULL;
-	RList/*<int>*/ *end_list = r_list_new ();
+	RList/*<int>*/ *end_list = r_list_newf(free);
```
Comment (crowell) [inline libr/core/cmd_search.c]: it's not an int any more, change the comment to say the type of struct

File: libr/core/cmd_search.c [hunk 7]
```diff
@@ -26,6 +26,12 @@ struct search_parameters {
 	boolt rsa_search;
 };
 
+struct endlist_pair {
+	intptr_t instr_offset;
+	int delay_size;
+};
+
+
```
Comment (crowell) [inline libr/core/cmd_search.c]: rm extra line

File: libr/core/cmd_search.c [hunk 8]
```diff
@@ -851,7 +864,19 @@ static int r_core_search_rop(RCore *core, ut64 from, ut64 to, int opt, const cha
 					// limit number of high level rop gadget results
 					break;
 				}
-				r_list_append (end_list, (void*)(intptr_t)i);
+				struct endlist_pair *epair = malloc(sizeof(struct endlist_pair));
+				// If this arch has branch delay slots, add the next instr as well
+				if ( end_gadget.delay && (end_gadget.type != R_ANAL_OP_TYPE_TRAP) ) {
```
Comment (crowell) [inline libr/core/cmd_search.c]: you can remove the trap thing, the op.delay will take care of that

File: libr/core/cmd_search.c [hunk 9]
```diff
@@ -658,6 +666,11 @@ static RList* construct_rop_gadget(RCore *core, ut64 addr, ut8 *buf, int idx,
 		r_list_append (badstart, p);
 	}
 	r_list_free (localbadstart);
+	//if our arch has bds then we better be including them
```
Comment (crowell) [inline libr/core/cmd_search.c]: space after `//`

File: libr/core/cmd_search.c (hunk 8 above)
Comment (crowell) [inline libr/core/cmd_search.c]: rm space after `(` on the `if ( end_`

**Comment by crowell** [comment]
thanks

**Pull request closed without merging**